docs/README.md

```diff
@@ -0,0 +1,134 @@
+
+## Getting started
+
+The types and names here are a fairly faithful representation of the components described in the spec.
+
+- [`URI`][URI] is for absolutely specified URIs that can also have path, query, and fragment (hash) parts.
+- [`AbsoluteURI`][AbsoluteURI] is a variation on `URI` that drops the ability for the URI to carry a fragment.
+- [`RelativeRef`][RelativeRef] is for relatively specified URIs that can also have path, query, and fragment (hash) parts.
+- [`URIRef`][URIRef] is combination of `URI` and `RelativeRef`, allowing the full range of representable URIs.
+
+The absolute/relative terminology when applied to URIs does not relate to the paths that a URI may carry, it refers to whether the URI has a "scheme" or not. For example `http://example.com` and `file://../test.txt` are absolute URIs but `//example.com` and `/test.txt` are relative.
+
+Assuming none of the `unsafe`-prefixed functions are used when constructing a URI, it should be impossible to construct a URI that is invalid using the types this library provides*. The slight downside of this is the data structures are relatively complex so as to only admit correct possibilities.
+
+\* Actually, there is one exception to that - `IPv6Address` is far too forgiving in what it allows currently. Contributions welcome!
+
+### URI component representations
+
+Due to the differing needs of users of this library, the URI types are all parameterised to allow for custom representations to be used for parts of the URI. Take a look at the most heavily parametrised type, `URIRef`:
+
+``` purescript
+type URIRef userInfo hosts path hierPath relPath query fragment = ...
+```
+
+This allows us to provide hooks into the parsing and printing processes for a URI, so that types better suited to the intended use case can be used.
+
+Taking `userInfo` as an example, according to the spec, the `user-info` part of an authority is just an arbitrary string of characters terminated by an `@` before a hostname. An extremely common usage for this is the `user:password` scheme, so by leaving the choice of representation as a type variable we can switch it out for a type specifically designed to handle that (this library includes one actually, under [`URI.Extra.UserPassInfo`][UserPassInfo]).
+
+### App-specific URI type definitions
+
+When using this library, you'll probably want to define type synonyms for the URIs that make sense for your use case. A URI type that uses the simple representations for each component will look something like this:
+
+``` purescript
+type MyURI = URIRef UserInfo (HostPortPair Host Port) Path HierPath RelPath Query Fragment
+```
+
+Along with these types, you'll want to define an options record that specifies how to parse and print URIs that look like this:
+
+``` purescript
+options ∷ Record (URIRefOptions UserInfo (HostPortPair Host Port) Path HierPath RelPath Query Fragment)
```

Comment:
This hasn't been mentioned before, but I would like to standardize on no unicode symbols within contrib. I think they're perfectly fine in personal code, but it certainly seems standard in the community to use non-unicode characters (`::` instead of `∷` here) and it means cotributors who don't use them don't have to learn how to create or copy-paste them. If you have a moment to glance through and replace them in the docs I think that would be useful.

Reply:
I think every place you would expect `::`, Gary used `∷`. I think I prefer the unicode character because things line up better and `forall` is reduced to something much smaller. However, I realize that could confuse new people who aren't sure what that means.
Regardless, I think we should address this in a separate PR.